Raise EBADF when reading or writing to already closed file.

mdboom opened the pull request:
This is another thing found by running the Python test suite.
Without this check, it ends up going into the pipes code where `pipes.buffer` has already been set to `null` when the pipe was closed.

kripken commented on src/library_fs.js:
```diff
@@ -1100,6 +1101,9 @@ mergeInto(LibraryManager.library, {
       if (length < 0 || position < 0) {
         throw new FS.ErrnoError(ERRNO_CODES.EINVAL);
       }
+      if (stream.fd === null) {
```
how about adding a helper function, `isClosed(stream)` that would check this? then it would be self-explanatory, otherwise it's not clear what the `fd` being `null` implies.

Aside from that, lgtm with a test. Thanks!

kripken commented on tests/test_core.py:
```diff
@@ -4732,6 +4732,13 @@ def test_fs_write(self, js_engines=None):
     out = path_from_root('tests', 'fs', 'test_write.out')
     self.do_run_from_file(src, out, js_engines=js_engines)
 
+  @also_with_noderawfs
```
this should best be in `other` as in the other PR. as in there, please see #6473 for a good way to write tests there.

kripken replied:
Ah, unless we do want to test this with NODERAWFS?

kripken commented:
Hmm, the failures here seem closure compiler related, as if it is not minifying the name of the new method correctly. Is it possible it's used on an instance of something not of the right class - maybe it needs to be defined in another place?

kripken commented:
I think those errors can be fixed by changing `this.isClosed` to `FS.isClosed`. closure tries to identify the object, and its analysis depends on the code following certain conventions (I don't remember offhand the specific one that could make `this.` work here, but we use `FS.` anyhow).

mdboom commented:
Great.  I changed `this.isClosed` to `FS.isClosed` and we'll see how the CI likes it...

kripken commented:
Progress, looks like tests are green except for the new test added here.

mdboom commented:
Fixed the test, hopefully.

You learn something new everyday: I think the low-level file descriptor-based API is able to detect that a file descriptor is closed like this, but the higher-level `FILE *` API can't.  As far as I can tell, using a `FILE *` after it's closed is undefined.  `musl`, at least, `free`'s the FILE object in close, so all bets are off.  So I updated the test to use the low-level API (which is what the test in Python that lead to this PR was doing anyways).

kripken commented:
Interesting. The C API is weird...
